### PR TITLE
[SelectionDAG] Calculate KnownBits for SMIN

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

**Comment by topperc** [comment]
Tests? I feel like I keep asking for tests on many of your patches.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -4017,6 +4015,16 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
       }
     }
 
+    // For SMIN, if CstHigh is negative we know the result will be
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: I don't understand why any of this SMIN/SMAX handling code is needed, except for the call to KnownBits::smin/smax. Those functions are optimal, and the code here does not seem to use any knowledge that is not captured by KnownBits info.

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: It's possible to know the LHS has some number of sign bits without knowing whether it is positive or negative. Here's the patch that added the SMAX code https://reviews.llvm.org/D126896

Comment (jayfoad) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4018]: Thanks - I see now that using numsignbits info does let it deduce more stuff than KnownBits can.

**Review by arsenm — COMMENTED** [review]
This would be OK with tests

**Pull request closed without merging**